modules/core/src/main/java/org/apache/ignite/internal/processors/cache/consistentcut/ConsistentCut.java

```diff
@@ -0,0 +1,260 @@
+import org.apache.ignite.lang.IgniteUuid;
+import org.jetbrains.annotations.Nullable;
+
+import static org.apache.ignite.internal.processors.cache.transactions.IgniteInternalTx.FinalizationStatus.RECOVERY_FINISH;
+import static org.apache.ignite.transactions.TransactionState.ACTIVE;
+import static org.apache.ignite.transactions.TransactionState.COMMITTED;
+import static org.apache.ignite.transactions.TransactionState.MARKED_ROLLBACK;
+import static org.apache.ignite.transactions.TransactionState.ROLLED_BACK;
+import static org.apache.ignite.transactions.TransactionState.UNKNOWN;
+
+/**
+ * Describes current Consistent Cut.
+ */
+public class ConsistentCut extends GridFutureAdapter<WALPointer> {
+    /** */
+    private final GridCacheSharedContext<?, ?> cctx;
+
+    /** */
+    private final IgniteLogger log;
+
+    /** ID of Consistent Cut. */
+    private final UUID id;
+
+    /** Set of checked transactions belonging to the BEFORE side. */
+    @GridToStringInclude
+    private Set<GridCacheVersion> beforeCut;
+
+    /** Set of checked transactions belonging to the AFTER side. */
+    @GridToStringInclude
+    private Set<GridCacheVersion> afterCut;
+
+    /** Collection of transactions removed from {@link IgniteTxManager#activeTransactions()}. */
+    private volatile Set<IgniteInternalFuture<IgniteInternalTx>> removedActive = ConcurrentHashMap.newKeySet();
+
+    /** */
+    ConsistentCut(GridCacheSharedContext<?, ?> cctx, UUID id) {
+        this.cctx = cctx;
+        this.id = id;
+
+        log = cctx.logger(ConsistentCut.class);
+    }
+
+    /** */
+    public UUID id() {
+        return id;
+    }
+
+    /**
+     * Inits local Consistent Cut: prepares list of active transactions to check which side of Consistent Cut they belong to.
+     */
+    protected void init() throws IgniteCheckedException {
+        walLog(new ConsistentCutStartRecord(id), false);
+
+        beforeCut = ConcurrentHashMap.newKeySet();
+        afterCut = ConcurrentHashMap.newKeySet();
+
+        GridCompoundFuture<Boolean, Boolean> checkFut = new GridCompoundFuture<>(CU.boolReducer());
+
+        Iterator<IgniteInternalFuture<IgniteInternalTx>> finFutIt = cctx.tm().activeTransactions().stream()
+            .filter(tx -> tx.state() != ACTIVE)
+            .map(IgniteInternalTx::finishFuture)
+            .iterator();
+
+        // Invoke sequentially over two iterators:
+        // 1. iterators are weakly consistent.
+        // 2. we need a guarantee to handle `removedActive` after `activeTxs` to avoid missed transactions.
+        checkTransactions(finFutIt, checkFut);
+        checkTransactions(removedActive.iterator(), checkFut);
+
+        removedActive = null;
+
+        checkFut.markInitialized();
+
+        checkFut.listen(finish -> {
+            if (Boolean.FALSE.equals(finish.result()) || isDone()) {
+                if (log.isDebugEnabled())
+                    log.debug("Cut might be inconsistent for id " + id + ". Skip writing FinishRecord.");
```

Comment:
Why "might"? Is there any case when finish.result == FALSE and cut is consistent?

Let's log

Reply:
Actually when finish.result == FALSE, Ignite node just doesn't have enough info to know whether this cut is consistent or not. Then we just consider it inconsistent and skip writing the FinishRecord.